### PR TITLE
Add week_of_year Spark function

### PR DESCRIPTION
**Pull request by Ma-Jian1**
Presto's `week_of_year` function differs from Spark's in return type: BIGINT vs. INTEGER.

Copy-paste Presto's implementation and change the return type.

Spark's implementation: https://github.com/apache/spark/blob/v3.2.2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L868

### CODE REVIEW

**Comment by Ma-Jian1** [comment]
@rui-mo

File: velox/docs/functions/spark/datetime.rst [hunk 1]
```diff
@@ -55,3 +61,7 @@ These functions support TIMESTAMP and DATE input types.
 .. spark:function:: year(x) -> integer
 
     Returns the year from ``x``.
+    
+.. spark:function:: add_months(startDate, numMonths) -> integer
```
Comment (rui-mo) [inline velox/docs/functions/spark/datetime.rst:65]: This part should belong to another PR, right?

File: velox/functions/sparksql/Register.cpp [hunk 2]
```diff
@@ -203,6 +205,8 @@ void registerFunctions(const std::string& prefix) {
       {prefix + "make_date"});
 
   registerFunction<LastDayFunction, Date, Date>({prefix + "last_day"});
+  registerFunction<AddMonthsFunction, Date, Date, int32_t>(
+      {prefix + "add_months"});
```
Comment (rui-mo) [inline velox/functions/sparksql/Register.cpp:209]: Same.

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp [hunk 3]
```diff
@@ -198,5 +224,25 @@ TEST_F(DateTimeFunctionsTest, lastDay) {
   EXPECT_EQ(lastDayFunc(std::nullopt), std::nullopt);
 }
 
+TEST_F(DateTimeFunctionsTest, addMonths) {
```
Comment (rui-mo) [inline velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:227]: Same.

**Review by rui-mo — COMMENTED** [review]
Can we put Spark's implementation to the PR description?

**Comment by Ma-Jian1** [comment]
I0807 10:11:07.190543 1754778 ExpressionFuzzer.cpp:1313] ==============================> Started iteration 74530 (seed: 1635755765)
I0807 10:11:07.344303 1754778 ExpressionVerifier.cpp:91] Executing expression 0 : week_of_year("c0")
I0807 10:11:07.344316 1754778 ExpressionVerifier.cpp:31] 1 vectors as input:
I0807 10:11:07.344318 1754778 ExpressionVerifier.cpp:33]        [FLAT TIMESTAMP: 100 elements, 12 nulls]
I0807 10:11:07.344692 1754778 ExpressionVerifier.cpp:79] All results match.
I0807 10:11:07.344715 1754778 ExpressionFuzzer.cpp:1358] ==============================> Done with iteration 74530
I0807 10:11:07.344756 1754778 ExpressionFuzzer.cpp:1093] ==============================> Top 1 by number of rows processed
I0807 10:11:07.344761 1754778 ExpressionFuzzer.cpp:1095] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0807 10:11:07.344763 1754778 ExpressionFuzzer.cpp:1099] week_of_year 74531 100.00% 1142417
I0807 10:11:07.344866 1754778 ExpressionFuzzer.cpp:1105] ==============================> Bottom 1 by number of rows processed
I0807 10:11:07.344869 1754778 ExpressionFuzzer.cpp:1107] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0807 10:11:07.344873 1754778 ExpressionFuzzer.cpp:1112] week_of_year 74531 100.00% 1142417
I0807 10:11:07.344879 1754778 ExpressionFuzzer.cpp:1125] ==============================> All stats sorted by number of times the function was chosen
I0807 10:11:07.344884 1754778 ExpressionFuzzer.cpp:1127] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0807 10:11:07.344889 1754778 ExpressionFuzzer.cpp:1131] week_of_year 74531 100.00% 1142417
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp [hunk 4]
```diff
@@ -66,6 +70,26 @@ TEST_F(DateTimeFunctionsTest, yearDate) {
   EXPECT_EQ(1920, year(DATE()->toDays("1920-01-01")));
 }
 
+TEST_F(DateTimeFunctionsTest, weekDate) {
```
Comment (rui-mo) [inline velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:73]: When the input is invalid date, is the behavior the same as Spark's?

File: velox/docs/functions/spark/datetime.rst [hunk 5]
```diff
@@ -22,6 +22,12 @@ These functions support TIMESTAMP and DATE input types.
     ``day`` need to be from 1 to 31, and matches the number of days in each month.
     days of ``year-month-day - 1970-01-01`` need to be in the range of INTEGER type.
 
+.. function:: week(x) -> integer
+
+    Returns the `ISO-Week`_ of the year from x. The value ranges from ``1`` to ``53``.
```
Comment (rui-mo) [inline velox/docs/functions/spark/datetime.rst:27]: In Spark, `A week is considered to start on a Monday and week 1 is the first week with >3 days.`
Can we document this also? And is current implementation the same with above statement?

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp (hunk 4 above)
Comment (Ma-Jian1) [reply velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:73]: spark just test null for invalid input, which can't be tested in velox, no other exception test.

File: velox/docs/functions/spark/datetime.rst (hunk 5 above)
Comment (Ma-Jian1) [reply velox/docs/functions/spark/datetime.rst:27]: `A week is considered to start on a Monday and week 1 is the first week with >3 days.` this is described in `ISO-Week`.
It's covered by the ut. `1919-12-31` => `1`; `1920-01-01` => `1`

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp (hunk 4 above)
Comment (rui-mo) [reply velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:73]: What about `1997-02-29` which is invalid?

File: velox/docs/functions/spark/datetime.rst (hunk 5 above)
Comment (rui-mo) [reply velox/docs/functions/spark/datetime.rst:27]: Got it. Can we record this in the doc?

> A week is considered to start on a Monday and week 1 is the first week with >3 days.

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp (hunk 4 above)
Comment (Ma-Jian1) [reply velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:73]: It's coverd by velox function ` daysSinceEpochFromDate`
```cpp
C++ exception with description "Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: Date out of range: 1997-2-29
Retriable: False
Function: daysSinceEpochFromDate
File: ../../velox/type/TimestampConversion.cpp
Line: 455
Stack trace:
```

**Comment by Ma-Jian1** [comment]
@mbasmanova Could you help to review this pr?

**Review by mbasmanova — COMMENTED** [review]
@Ma-Jian1 Would you review contributing guide and update accordingly. Specifically, make sure the PR title and description are in compliance.

**Comment by Ma-Jian1** [comment]
> @Ma-Jian1 Would you review contributing guide and update accordingly. Specifically, make sure the PR title and description are in compliance.

Sorry for that.
Update PR title and document according to the guide.

File: velox/docs/functions/spark/datetime.rst [hunk 6]
```diff
@@ -52,6 +52,13 @@ These functions support TIMESTAMP and DATE input types.
     Returns null if ``string`` does not match ``format`` or if ``format``
     is invalid.
 
+.. function:: week(x) -> integer
```
Comment (mbasmanova) [inline velox/docs/functions/spark/datetime.rst:55]: The PR title says "week_of_year", but this doc says "week". Which is correct?

File: velox/docs/functions/spark/datetime.rst [hunk 7]
```diff
@@ -52,6 +52,13 @@ These functions support TIMESTAMP and DATE input types.
...
+
+    Returns the `ISO-Week`_ of the year from x. The value ranges from ``1`` to ``53``.
+    A week is considered to start on a Monday and week 1 is the first week with >3 days.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/datetime.rst:58]: Would you double check that this matches Spark's implementation?

File: velox/docs/functions/spark/datetime.rst (hunk 6 above)
Comment (Ma-Jian1) [reply velox/docs/functions/spark/datetime.rst:55]: fixed to `week_of_year`

File: velox/docs/functions/spark/datetime.rst (hunk 7 above)
Comment (Ma-Jian1) [reply velox/docs/functions/spark/datetime.rst:58]: Add some extra ut that copy from Spark UT, and validate the results.
From https://github.com/apache/spark/blob/v3.2.2/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L3240-L3241, It also follows ISO 8601

**Review by mbasmanova — APPROVED** [review]
Thanks. There are some tooling issues blocking me from merging this PR. These should be resolved sometime tomorrow.

**Comment by Ma-Jian1** [comment]
@mbasmanova Can this PR be merged?

**Comment by mbasmanova** [comment]
Let me try. Would you rebase?
